crates/build-rs/src/output.rs

```diff
@@ -0,0 +1,408 @@
+    };
+    if path.contains('\n') {
+        panic!("cannot emit rustc-link-search: path contains newline");
+    }
+    emit("rustc-link-search", format_args!("{kind}={path}"));
+}
+
+/// The `rustc-flags` instruction tells Cargo to pass the given space-separated
+/// flags to the compiler. This only allows the `-l` and `-L` flags, and is
+/// equivalent to using [`rustc_link_lib`] and [`rustc_link_search`].
+#[track_caller]
+pub fn rustc_flags(flags: &str) {
+    if flags.contains('\n') {
+        panic!("cannot emit rustc-flags: invalid flags");
+    }
+    emit("rustc-flags", flags);
+}
+
+/// The `rustc-cfg` instruction tells Cargo to pass the given value to the
+/// [`--cfg` flag][cfg] to the compiler. This may be used for compile-time
+/// detection of features to enable conditional compilation.
+///
+/// Note that this does not affect Cargo’s dependency resolution. This cannot
+/// be used to enable an optional dependency, or enable other Cargo features.
+///
+/// Be aware that [Cargo features] use the form `feature="foo"`. `cfg` values
+/// passed with this flag are not restricted to that form, and may provide just
+/// a single identifier, or any arbitrary key/value pair. For example, emitting
+/// `rustc_cfg("abc")` will then allow code to use `#[cfg(abc)]` (note the lack
+/// of `feature=`). Or an arbitrary key/value pair may be used with an `=` symbol
+/// like `rustc_cfg(r#"my_component="foo""#)`. The key should be a Rust identifier,
+/// the value should be a string.
+///
+/// [cfg]: https://doc.rust-lang.org/rustc/command-line-arguments.html#option-cfg
+/// [Cargo features]: https://doc.rust-lang.org/cargo/reference/features.html
+#[track_caller]
+pub fn rustc_cfg(key: &str) {
+    if !is_ident(key) {
+        panic!("cannot emit rustc-cfg: invalid key {key:?}");
+    }
+    emit("rustc-cfg", key);
+}
+
+/// Like [`rustc_cfg`], but with the value specified separately. To replace the
+/// less convenient `rustc_cfg(r#"my_component="foo""#)`, you can instead use
+/// `rustc_cfg_value("my_component", "foo")`.
+#[track_caller]
+pub fn rustc_cfg_value(key: &str, value: &str) {
+    if !is_ident(key) {
+        panic!("cannot emit rustc-cfg-value: invalid key");
+    }
+    let value = value.escape_default();
+    emit("rustc-cfg", format_args!("{key}=\"{value}\""));
+}
+
+/// Add to the list of expected config names that is used when checking the
+/// *reachable* cfg expressions with the [`unexpected_cfgs`] lint.
+///
+/// This form is for keys without an expected value, such as `cfg(name)`.
+///
+/// It is recommended to group the `rustc_check_cfg` and `rustc_cfg` calls as
+/// closely as possible in order to avoid typos, missing check_cfg, stale cfgs,
+/// and other mistakes.
+///
+/// [`unexpected_cfgs`]: https://doc.rust-lang.org/rustc/lints/listing/warn-by-default.html#unexpected-cfgs
+#[doc = msrv!("1.80")]
+#[track_caller]
+pub fn rustc_check_cfgs(keys: &[&str]) {
```

Comment:
`rustc-check-cfg` is the most "interesting" API to translate, and I'm the least happy with the way it's exposed here. 

The split between `rustc_cfg` and `rustc_cfg_value` isn't *great*, but it's at least straightforward to translate optional arguments to multiple function names. 

I'm like 20% tempted to expose this as a macro so that the directive syntax for 1+ keys and 0+ values can be directly replicated here. But it's been an explicit goal for this bridge to expose *simple*, straightforward API, and macro tricks are anything but. 

Sketch, without the necessary macro trickery:

```rust
macro_rules! rustc_check_cfg {
    ($($key:expr),+) => { rustc_check_cfg(&[$($key),+]) };
    ($($key:expr),+, values()) => { rustc_check_cfg_value(&[$($key),+], &[]) };
    ($($key:expr),+, values(none())) => { rustc_check_cfg(&[$($key),+]) };
    ($($key:expr),+, values(any())) => { rustc_check_cfg_any(&[$($key),+]) };
    ($($key:expr),+, values($($value:expr),*)) => { rustc_check_cfg_value(&[$($key),+], &[$($value),*) };
}
```

Other options: 
- Trait for accepting `&str` or `&[&str]`. Still a bespoke API generalization that makes for worse type errors/inference.
- Builder style API. Requires either `.emit()` or printing on drop.

Reply:
While code feedback is good, we should probably decouple it from this PR and handle it in follow ups.